source/_posts/2018-02-26-URLComponents.md

```diff
@@ -0,0 +1,367 @@
+        return parametersAfterParameterName
+    }
+
+    //cut the substring where the next parameter starts
+    return String(parametersAfterParameterName.prefix(upTo: rangeOfNextAmpersand))
+}
+
+//test
+if let url = URL(string: "https://showsknownfrom.tv/shows/12345678?featured=true") {
+    if let parameter = receivedUrl(url, contains: "featured") {
+        print("found \(parameter)") //gives us "found true"
+    } else {
+        print("not found")
+    }
+}
+```
+
+Granted, it works but...wow! So much going on here, magic values and String gymnastics galore.
+
+We could probably make this prettier but...why should we, when something as beautiful as `URLComponents` exists.
+
+## Composing and Decomposing URLs - The Easy Way
+By now you should be more than ready for `URLComponents` so lets dive in.
+
+### URLComponents - An Introduction
+`URLComponents` was introduced in iOS 7.0 and macOS 10.9 so it has been around for some time.
+
+To create a `URLComponents` object you can use either a `String` or a `URL`, in both cases you'll end out with an optional `URLComponents` object.
+
+Just a quick note about the `init` method if you are using a `URL`. As it says in the [documentation](https://developer.apple.com/documentation/foundation/urlcomponents/1780039-init):
+
+> If resolvingAgainstBaseURL is true and url is a relative URL, the components of url.absoluteURL are used. If the url string from the URL is malformed, nil is returned.
+
+To create a new `URLComponents` object we can write something like this:
+
+```swift
+let urlString = "https://showsknownfrom.tv/search?q=Sopranos&order=desc"
+
+guard var urlComponents = URLComponents(string: urlString) else {
+    return nil
+}
+```
+
+Now we have access to all parts of the `URL` and can probe them as we see fit.
+
+Here are some examples:
+
+```
+urlComponents.host   //gives us: showsknownfrom.tv
+urlComponents.scheme //gives us: https
+urlComponents.query  //gives us: q=Sopranos&order=desc
+```
+
+But the best part is this one:
+
+```swift
+urlComponents.queryItems
+```
+
+Which gives us back an array of `URLQueryItem` objects, where `URLQueryItem` objects are basically just key/value objects for the individual query items (documented [here](https://developer.apple.com/documentation/foundation/urlcomponents/1779966-queryitems)).
+
+Oh joy! That makes it so much easier for us to append query parameters, or check if a URL contains a query parameter.
+
+### Composing URLs for Searching
+Armed with our new knowledge, lets now look at how we can use `URLComponents` to build a `URL` with parameters for us:
+
+```swift
+func searchURL(with q: String, optionalParameters: [String: String]? = nil) -> URL? {
+    let urlString = "https://showsknownfrom.tv/search"
+
+    guard var urlComponents = URLComponents(string: urlString) else {
+        return nil
+    }
+
+    var queryItems: [URLQueryItem] = [URLQueryItem(name: "q", value: q)]
+
+    if let optionalParameters = optionalParameters {
```

Comment:
You could do something similar as my suggestion is the non-URLQueryItems approach above